Parallel cmdstanr

rok-cesnovar opened the pull request:
I started working on the parallel cmdstanr issue and before I dive to deep I wanted to get some feedback.
What is done at the moment:
  - runs parallel chains
  - outputs the iteration progress only 
  - sample is still a blocking function

What is missing:
  - parsing other output than iteration (warnings, errors, etc)
  - non-blocking variant

I think sampling should be a blocking function by default, otherwise its going to be confusing for the users. But that is just my opinion, open to suggestions.

rok-cesnovar commented:
Example of the current output:

```
Running ./lr_glm 'id=1' random 'seed=653079466' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/RtmpCUwajV/lr_glm-stan-sample-1.csv' 'method=sample' \
  'num_samples=300' 'num_warmup=300' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=2' random 'seed=720917795' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/RtmpCUwajV/lr_glm-stan-sample-2.csv' 'method=sample' \
  'num_samples=300' 'num_warmup=300' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=3' random 'seed=1349904954' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/RtmpCUwajV/lr_glm-stan-sample-3.csv' 'method=sample' \
  'num_samples=300' 'num_warmup=300' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=4' random 'seed=1530272637' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/RtmpCUwajV/lr_glm-stan-sample-4.csv' 'method=sample' \
  'num_samples=300' 'num_warmup=300' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
CHAIN 1: Iteration:   1 / 600 [  0%]  (Warmup)
CHAIN 2: Iteration:   1 / 600 [  0%]  (Warmup)
CHAIN 1: Iteration: 100 / 600 [ 16%]  (Warmup)
CHAIN 3: Iteration:   1 / 600 [  0%]  (Warmup)
CHAIN 4: Iteration:   1 / 600 [  0%]  (Warmup)
CHAIN 3: Iteration: 100 / 600 [ 16%]  (Warmup)
CHAIN 2: Iteration: 100 / 600 [ 16%]  (Warmup)
CHAIN 4: Iteration: 100 / 600 [ 16%]  (Warmup)
CHAIN 1: Iteration: 200 / 600 [ 33%]  (Warmup)
CHAIN 3: Iteration: 200 / 600 [ 33%]  (Warmup)
CHAIN 2: Iteration: 200 / 600 [ 33%]  (Warmup)
CHAIN 1: Iteration: 300 / 600 [ 50%]  (Warmup)
CHAIN 1: Iteration: 301 / 600 [ 50%]  (Sampling)
CHAIN 3: Iteration: 300 / 600 [ 50%]  (Warmup)
CHAIN 2: Iteration: 300 / 600 [ 50%]  (Warmup)
CHAIN 3: Iteration: 301 / 600 [ 50%]  (Sampling)
CHAIN 2: Iteration: 301 / 600 [ 50%]  (Sampling)
CHAIN 1: Iteration: 400 / 600 [ 66%]  (Sampling)
CHAIN 2: Iteration: 400 / 600 [ 66%]  (Sampling)
CHAIN 4: Iteration: 200 / 600 [ 33%]  (Warmup)
CHAIN 1: Iteration: 500 / 600 [ 83%]  (Sampling)
CHAIN 3: Iteration: 400 / 600 [ 66%]  (Sampling)
CHAIN 2: Iteration: 500 / 600 [ 83%]  (Sampling)
CHAIN 1: Iteration: 600 / 600 [100%]  (Sampling)
CHAIN 4: Iteration: 300 / 600 [ 50%]  (Warmup)
CHAIN 4: Iteration: 301 / 600 [ 50%]  (Sampling)
CHAIN 3: Iteration: 500 / 600 [ 83%]  (Sampling)
CHAIN 2: Iteration: 600 / 600 [100%]  (Sampling)
CHAIN 4: Iteration: 400 / 600 [ 66%]  (Sampling)
CHAIN 3: Iteration: 600 / 600 [100%]  (Sampling)
CHAIN 4: Iteration: 500 / 600 [ 83%]  (Sampling)
CHAIN 4: Iteration: 600 / 600 [100%]  (Sampling)
```

jgabry commented:
Thanks for working on this! Good point that not blocking could be confusing for many users. Maybe at some point in the future we can make it optional for people who want it, but we can decide whether to do that another time. This looks good so far!

bbbales2 commented:
> sample is still a blocking function

I agree with that being the default behavior

> parsing other output than iteration (warnings, errors, etc)
> non-blocking variant

Neither of these are necessary for me to find this very useful, so I'm happy if those wait until other pulls.

If possible I would like the Rstudio stop button to stop all the chains right out of the gate. I remember for a long time with Rstan this was an issue (and if you stopped a job you'd have to go kill the chains manually).

I'm happy to test the functionality. Just let me know when.

bbbales2 commented:
Also thanks for doing this!

rok-cesnovar commented:
> If possible I would like the Rstudio stop button to stop all the chains right out of the gate.

100% agree on this. The version currently pushed doesnt have this, but I did fix it so that the processes are stopped. Just need to polish a few more stuff and then will ping you. Thanks!

rok-cesnovar commented:
I am removing the [WIP] as this is ready for a proper look. Will go for blocking sampling with this PR.

Current behaviour of the output parser:
- outputs everything after the line "Adjust your expectations accordingly!". If anyone is interested in the start info (arguments and such) they can print that afterwards with fit$output().
- empty lines are not printed
- not outputing the total time from cmdstan, instead using Sys.time and printing the mean at the end (its more useful and realistic)
- fit$time() returns a vector of execution times. is that fine?
- fit$output(chain_id) prints the entire output for a chain. Do we return the vector of lines here? Not sure if parsing that is a use case. Probably not.

Example output:
```
Running MCMC sampling with 4 chain(s)...
Running ./lr_glm 'id=1' random 'seed=1589220072' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/Rtmpc0AszH/lr_glm-stan-sample-1.csv' 'method=sample' \
  'num_samples=500' 'num_warmup=500' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=2' random 'seed=850416257' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/Rtmpc0AszH/lr_glm-stan-sample-2.csv' 'method=sample' \
  'num_samples=500' 'num_warmup=500' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=3' random 'seed=247236796' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/Rtmpc0AszH/lr_glm-stan-sample-3.csv' 'method=sample' \
  'num_samples=500' 'num_warmup=500' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
Running ./lr_glm 'id=4' random 'seed=659707608' data 'file=/home/rok/Desktop/scripts/test.json' output 'file=/tmp/Rtmpc0AszH/lr_glm-stan-sample-4.csv' 'method=sample' \
  'num_samples=500' 'num_warmup=500' 'save_warmup=0' 'algorithm=hmc' 'engine=nuts' adapt 'engaged=1'
CHAIN 1: Iteration:   1 / 1000 [  0%]  (Warmup)
CHAIN 2: Iteration:   1 / 1000 [  0%]  (Warmup)
CHAIN 3: Iteration:   1 / 1000 [  0%]  (Warmup)
CHAIN 4: Iteration:   1 / 1000 [  0%]  (Warmup)
CHAIN 1: Iteration: 100 / 1000 [ 10%]  (Warmup)
CHAIN 3: Iteration: 100 / 1000 [ 10%]  (Warmup)
CHAIN 1: Iteration: 200 / 1000 [ 20%]  (Warmup)
CHAIN 2: Iteration: 100 / 1000 [ 10%]  (Warmup)
CHAIN 3: Iteration: 200 / 1000 [ 20%]  (Warmup)
CHAIN 4: Iteration: 100 / 1000 [ 10%]  (Warmup)
CHAIN 1: Iteration: 300 / 1000 [ 30%]  (Warmup)
CHAIN 3: Iteration: 300 / 1000 [ 30%]  (Warmup)
CHAIN 1: Iteration: 400 / 1000 [ 40%]  (Warmup)
CHAIN 3: Iteration: 400 / 1000 [ 40%]  (Warmup)
CHAIN 4: Iteration: 200 / 1000 [ 20%]  (Warmup)
CHAIN 2: Iteration: 200 / 1000 [ 20%]  (Warmup)
CHAIN 1: Iteration: 500 / 1000 [ 50%]  (Warmup)
CHAIN 1: Iteration: 501 / 1000 [ 50%]  (Sampling)
CHAIN 3: Iteration: 500 / 1000 [ 50%]  (Warmup)
CHAIN 3: Iteration: 501 / 1000 [ 50%]  (Sampling)
CHAIN 1: Iteration: 600 / 1000 [ 60%]  (Sampling)
CHAIN 4: Iteration: 300 / 1000 [ 30%]  (Warmup)
CHAIN 2: Iteration: 300 / 1000 [ 30%]  (Warmup)
CHAIN 3: Iteration: 600 / 1000 [ 60%]  (Sampling)
CHAIN 4: Iteration: 400 / 1000 [ 40%]  (Warmup)
CHAIN 2: Iteration: 400 / 1000 [ 40%]  (Warmup)
CHAIN 1: Iteration: 700 / 1000 [ 70%]  (Sampling)
CHAIN 4: Iteration: 500 / 1000 [ 50%]  (Warmup)
CHAIN 4: Iteration: 501 / 1000 [ 50%]  (Sampling)
CHAIN 2: Iteration: 500 / 1000 [ 50%]  (Warmup)
CHAIN 2: Iteration: 501 / 1000 [ 50%]  (Sampling)
CHAIN 3: Iteration: 700 / 1000 [ 70%]  (Sampling)
CHAIN 4: Iteration: 600 / 1000 [ 60%]  (Sampling)
CHAIN 1: Iteration: 800 / 1000 [ 80%]  (Sampling)
CHAIN 2: Iteration: 600 / 1000 [ 60%]  (Sampling)
CHAIN 3: Iteration: 800 / 1000 [ 80%]  (Sampling)
CHAIN 4: Iteration: 700 / 1000 [ 70%]  (Sampling)
CHAIN 2: Iteration: 700 / 1000 [ 70%]  (Sampling)
CHAIN 1: Iteration: 900 / 1000 [ 90%]  (Sampling)
CHAIN 4: Iteration: 800 / 1000 [ 80%]  (Sampling)
CHAIN 2: Iteration: 800 / 1000 [ 80%]  (Sampling)
CHAIN 3: Iteration: 900 / 1000 [ 90%]  (Sampling)
CHAIN 4: Iteration: 900 / 1000 [ 90%]  (Sampling)
CHAIN 1: Iteration: 1000 / 1000 [100%]  (Sampling)
CHAIN 2: Iteration: 900 / 1000 [ 90%]  (Sampling)
CHAIN 3: Iteration: 1000 / 1000 [100%]  (Sampling)
CHAIN 4: Iteration: 1000 / 1000 [100%]  (Sampling)
CHAIN 2: Iteration: 1000 / 1000 [100%]  (Sampling)
All chains have finished.
Average execution time of the 4 chains: 2.4 seconds
```
cc @bbbales2 : Do your thing :) and thanks

rok-cesnovar commented:
In case of a runtime critical exception it will end like this: 

```
CHAIN 1: Exception: mismatch in dimension declared and found in context; processing stage=data initialization; variable name=X; position=1; dims declared=(20000,11); dims found=(20000,10)  (in '/home/rok/Desktop/scripts/lr_glm.stan' at line 4)
CHAIN 2: Exception: mismatch in dimension declared and found in context; processing stage=data initialization; variable name=X; position=1; dims declared=(20000,11); dims found=(20000,10)  (in '/home/rok/Desktop/scripts/lr_glm.stan' at line 4)
CHAIN 4: Exception: mismatch in dimension declared and found in context; processing stage=data initialization; variable name=X; position=1; dims declared=(20000,11); dims found=(20000,10)  (in '/home/rok/Desktop/scripts/lr_glm.stan' at line 4)
CHAIN 3: Exception: mismatch in dimension declared and found in context; processing stage=data initialization; variable name=X; position=1; dims declared=(20000,11); dims found=(20000,10)  (in '/home/rok/Desktop/scripts/lr_glm.stan' at line 4)
All chains were stopped by critical exceptions.
```

bbbales2 commented:
Works great! I tested stop buttons and a model with errors and verified that output and diagnostics were still getting written correctly.

The only important comments I have are update the docs in model-method-sample for num_chains to indicate things run in parallel now. Is there a way to specify the maximum number of cores to use?

Second I made a model where some chains failed and some succeeded (run like 10 of these and you'll probably get at least one that will fail):

```                                                                                                                                                                
transformed data {
  int run = 0;
  int b = bernoulli_rng(0.5);

  if(b > 0) {
    run = 1;
  }
}

parameters {
  vector[100] mu;
}

model {
  if(run == 1) {
    mu ~ normal(0, 1);
  } else {
    mu ~ normal(1 / 0.0, 1);
  }
}
```

I get back at the end:

```
Error in runset$time() : (list) object cannot be coerced to type 'double'
```

after the rest of the chains finished. This seems like it should be recoverable in some way. I don't think rerun the chains, but the partial output should be available.

I didn't actually review code. I'll leave that to @jgabry since I'm too lazy to learn R well enough to check a complicated issue like this.

Minor comments which you can ignore if you want:

1. fit$output() should return a list with all the chains maybe? I dunno, empty is weird though.

3. Re: returning time, rstan's get_elapsed_time returns a little matrix:

```
        warmup  sample
chain:1  1e-06 1.06196
chain:2  1e-06 1.10443
chain:3  0e+00 1.10278
chain:4  1e-06 1.08515
```

(in those calculations I'd turned warmup off -- usually that'd be a large number)

4. > not outputing the total time from cmdstan, instead using Sys.time and printing the mean at the end (its more useful and realistic)

I like that as chains finish they print the time like they do in Rstan. Even if we printed them at the same time, it's good to see the variation.

Looks awesome though! Excited to get this in.

rok-cesnovar commented:
Huge thanks @bbbales2 for the thorough test.

> update the docs in model-method-sample for num_chains to indicate things run in parallel now

Done. Thanks for spotting that.

> Is there a way to specify the maximum number of cores to use?

Not right now. For now we create separate processes for all the chains and let the OS handle it.
We could limit this by only spawning `num_cores` chains and only spawn new ones once the existing finish. But I think that will be a separate PR, this one already has enough new wrinkles on it, I think.

> I get back at the end:
  Error in runset$time() : (list) object cannot be coerced to type 'double'

Ah yeah, didnt think of that scenario. I now refactored how times are reported.
This now prints each chain's execution time immediately upon finish and mean for all chains when all chains finish.

```
CHAIN 3: Iteration: 1900 / 2000 [ 95%]  (Sampling)
CHAIN 4: Iteration: 1700 / 2000 [ 85%]  (Sampling)
CHAIN 3: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 3 finished in 4.2 seconds.
CHAIN 4: Iteration: 1800 / 2000 [ 90%]  (Sampling)
CHAIN 4: Iteration: 1900 / 2000 [ 95%]  (Sampling)
CHAIN 4: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 4 finished in 4.5 seconds.
All 4 chains finished succesfully.
Mean execution time: 4.2 seconds
```

If some chains finish unexpectedly it looks like this:
```
Chain 3 finished unexpectedly!
Chain 4 finished in 0.1 seconds.
2 chain(s) finished unexpectedly!
The remaining chains had a mean execution time of 0.1 seconds
```

If all chains crash we then get:
```
Chain 4 finished unexpectedly!
All chains finished unexpectedly!
```

fit$time() also now returns a data.frame. Example:
```
  chain_id warmup_time sampling_time total_time
1        1    3.204252      1.441625   4.751414
2        2    2.221134      2.002872   4.294625
3        3    2.193038      1.942544   4.225951
4        4    1.963494      1.771150   3.962974
```
total_time = warmup_time + sampling_time + I/O

I am still using Sys.time() and not parsing the time reported by cmdstan. The cmdstan one is unreliable with threading and OpenCL so am just using R for that.

bbbales2 commented:
Hmm, I hope I didn't open a can of worms here saying we should expect to get the partial output with failed chains.

Now fit$summary throws an error if you run it with a model where some of the chains failed. fit$output_files and fit$diagnostic_files also return filenames from the failed chains (so they're difficult to work with because you need to add checking for failed run logic manually).

I guess a few options from here are:
1. If any chain dies, kill all chains. I really don't like this because then we don't get to diagnose what happened (why the chain died -- what's happening with the working chains, etc.)

2. Modify all the helper routines to work with only the processes that successfully ran. I think this is appropriate if we throw warnings at the end ("Chain(s) 1, 4 not included in output because they failed to complete"). If we're throwing warnings, I guess we also add a quiet flag so that they can be turned off.

3. Add flags to helper routines so that by default they use everything (and error) but can be switched into a permissive mode or something to work with the processes that successfully ran.

Another thing, it looks like there's lots of repeated output:
```
Chain 1 finished unexpectedly!
Chain 2 finished unexpectedly!
Chain 1 finished unexpectedly!
Chain 2 finished unexpectedly!
Chain 1 finished unexpectedly!
Chain 2 finished unexpectedly!
CHAIN 4: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1 finished unexpectedly!
Chain 2 finished unexpectedly!
Chain 4 finished in 0.2 seconds.
2 chain(s) finished unexpectedly!
The remaining chains had a mean execution time of 0.2 seconds
```

Like, chains 1 and 2 repeatedly complain about crashing. I'd expect them to just do it once and then go quiet until the end.

rok-cesnovar commented:
> Hmm, I hope I didn't open a can of worms here saying we should expect to get the partial output with failed chains.

Nah, I probably just goofed something up at the last minute. Will take a look this evening what did I do this time :) The last time I checked summary worked, but I am now not sure if I doubled checked it at the end.

Option 2 is the way to go I think. Thats how I handled the time() return here. It wont return the time if the chain didnt complete normally.

> Another thing, it looks like there's lots of repeated output:

Thanks for spotting this. 

@bbbales2 a bevarage of choice on me the next time we meet :) Really appreciate all the thorough testing.

bbbales2 commented:
> Option 2 is the way to go I think.

Hey I was talking to @jgabry today and he pointed out that the failures in different chains thing will also be a problem with the current serial implementation (this is indeed the case).

We can work this out in this pull req. or open a separate issue and work it out there. Doesn't bother me.

> a bevarage of choice on me the next time we meet

You're the one doing the work. I'm the happy end user here!

jgabry commented on R/model.R:
```diff
@@ -367,19 +374,67 @@ sample_method <- function(data = NULL,
 }
 CmdStanModel$set("public", name = "sample", value = sample_method)
 
+any_chain_alive <-function(procs, runset) {
```
what do you think about making `$any_chain_alive(procs)` a method for the runset object? not a big deal either way, but seems more consistent to include it in runset with the other methods you added (e.g., `chain_state`). 

If you want to do this then I think you just need to change the while loop to 
```r
while(runset$any_chain_alive(procs))
```
and use `self` instead of `runset` inside the function.

rok-cesnovar replied:
Great idea. Will fix.

jgabry replied:
I’m not sure but I wonder if `procs` itself should also go in `runset`. What do you think?

rok-cesnovar replied:
Good call. Will finish up this PR tomorrow. Spending my Mondays and Tuesdays TAing till February which makes me less productive :)

jgabry commented:
Also I do think it would be good to allow the user to specify the number of cores, but I agree that adding that can be separate from this PR.

jgabry commented:
@rok-cesnovar I saw some more commits here. Is this ready for me to review again? Not trying to rush you! Just checking if I should wait before I dive into this again if you're planning on adding more commits.

rok-cesnovar commented:
What is pushed is ready but I have the num_cores also done now. Just have to go through it again. Will tag you when its ready. Probably in an hour or so.

jgabry commented:
Cool!

rok-cesnovar commented:
@jgabry should be ready now

what changed:
- added warnings if chains finish unexpectedly
- cleaned up reapiting messages
- moved all functions inside runset
- you can now specify "num_cores" which limits the number of background process running at once
- added total_time for the entire sampling run 
- you can get draws, run summary & diagnose even if some chains fail

rok-cesnovar commented:
I just did another small update, now I am done for now.

jgabry commented:
Awesome thanks @rok-cesnovar! I will try to review today. 

@bbbales2 I can review the internal R code, but if you have time to try this out again (to see if you're happy with the behavior or if you can break it) that would be great too!

jgabry commented:
@rok-cesnovar Just tried it and it's great, but it still seems to get messed up by `refresh=0`.

rok-cesnovar commented:
ah, forgot the refresh = 0.

rok-cesnovar commented:
Fixed now @jgabry. Though with refresh=0 warmup and sampling time will be 0.
```
$total_time
[1] 0.9655704

$chain_time
  chain_id warmup_time sampling_time total_time
1        1           0             0  0.8520458
2        2           0             0  0.8453074
3        3           0             0  0.9446013
4        4           0             0  0.8350391
```
We can parse it out of the outpu, but we would then measure time differently with refresh=0. And as I said with threading and OpenCL that time is unreliable.

jgabry commented:
@rok-cesnovar This is great and we could have merged the PR as is, but I got a little carried away and decided to create a new R6 class CmdStanProcs, which is created by RunSet. I took all the stuff you added to RunSet and put it in there. I think this keeps the code a bit more organized. I was playing around with it anyway and wanted to try it out. What do you think? If you don't like it then we can just revert my commit and merge what you had. 
 
I also added a warning when `$time()` is called after refresh=0 (and set the times that were 0 to NA in that case).

rok-cesnovar commented:
Thanks! Looks good to me.

rok-cesnovar commented:
I added a more thorough check for errors (checked all possible err() calls in cmdstan) , so this now also fixes #69 

None of the added error can actually happen through vanilla cmdstanr but better be safe.

jgabry commented:
@rok-cesnovar Cool, thanks for adding the better error checks. 

To be safe I changed the default `num_cores`: 

* If `num_cores` is not specified then the default is now to check if the option `"mc.cores"` exists (this is what RStan does). this option can be set for an entire R session by options(mc.cores=value)
* If  `num_cores` is not specified and the `"mc.cores"` option has not been set then the default is 1 core to be safe

Is that behavior for `num_cores` ok with you?

jgabry commented:
Also, just so I'm 100% sure, are these the correct descriptions for the different "states" you're using? 

0) queued
1) starting
2) warmup
3) sampling
4) writing csv and finishing
5) finished ok 
6) finished unexpectedly

jgabry commented:
I also did a bit more reorganization which I think makes the code a bit more modular:

* CmdStanRun takes CmdStanArgs and  CmdStanProcs as input
* CmdStanProcs takes `num_runs` and `num_cores` as input
* CmdStanRun gets new method `$run_cmdstan()`, which drastically reduces the code in the model.R file

This has the nice effect that now the code for the `sample`, `optimize` and `variational` methods looks basically the same:

```r
  cmdstan_args <- CmdStanArgs$new(...)
  cmdstan_procs <- CmdStanProcs$new(num_runs, num_cores) 
  runset <- CmdStanRun$new(cmdstan_args, cmdstan_procs)
  runset$run_cmdstan()
  CmdStanFit$new(runset) # replace Fit with MCMC, MLE, VB but otherwise all these lines are the same
```

rok-cesnovar commented:
Thanks for reorganizing this. Looks nicer and more maintainable now. The mc.cores default seems like a better solution also. Thanks for the help!

bbbales2 commented:
Hey sorry if this was waiting on me. Just checked it out. Two comments on the output:

1. output_files gives us all the output files for things that work and I like that. But it would still be good to access files of chains that failed for debugging. Can we add an ```included_failed = FALSE``` argument to output_files and diagnostic_files? Probably should add that flag with save_output_files and save_diagnostic_files as well -- though in this cause I kinda feel like the default should be the opposite.

2. diagnostic_files returns files from all chains, whether they succeeded or failed.

We can just do these in a separate issue if you want. The basic functionality is working.

rok-cesnovar commented:
I would handle that separately, so we dont stall other PRs that will build on top of this one. If @jgabry agrees. Will make an issue. 

I understand 1. but not sure about 2.. Would you prefer that diagnostic_files dont return for failed chains or are you just pointing out that these 2 cases are handled differently. I agree with included_failed = FALSE.

bbbales2 commented:
Cool beans!

diagnostic_files has MCMC diagnostic stuff in it. It's not for diagnosing why a chain errored out or anything. So I think treating it the same as output_files makes sense.

jgabry commented:
Thanks @bbbales2! 

> I would handle that separately, so we dont stall other PRs that will build on top of this one. If @jgabry agrees. Will make an issue.

@rok-cesnovar Sounds good. Will merge this now so we can start tackling these other things.
